### DISCUSSION TITLE
User gets logged in automatically when resetting password

### DISCUSSION THREAD

**Discussion by emrecoban** [supabase/supabase#14209] [Questions] [answered]
I have an application where I use Next.js for Supabase Authentication. In this application, when a user receives a password reset email and clicks on the link to enter a new password, they are redirected to the link **but they are automatically logged in without being able to change their password** due to session control in `<SupabaseProvider>` or layout.js. I have been trying to solve this problem for hours but I haven't found a solution yet.

I am really stuck, and tired. It shouldn't be that difficult.

layout.js:
```javascript
import { createServerComponentSupabaseClient } from '@supabase/auth-helpers-nextjs'
import { headers, cookies } from 'next/headers'
import { redirect } from 'next/navigation'

export default async function AuthLayout({ children }) {
    const supabase = createServerComponentSupabaseClient({
        headers,
        cookies,
    })
    const { data } = await supabase.auth.getSession()

    if (data.session) {
        redirect('/admin')
    }

    return children
}
```

SupabaseProvider:
```javascript
'use client'

import { createContext, useContext, useEffect, useState } from 'react'
import { createBrowserSupabaseClient } from '@supabase/auth-helpers-nextjs'
import { useRouter } from 'next/navigation'

const Context = createContext(undefined)

export default function SupabaseProvider({ children }) {
    const [supabase] = useState(() => createBrowserSupabaseClient())
    const router = useRouter()

    useEffect(() => {
        const {
            data: { subscription },
        } = supabase.auth.onAuthStateChange((event) => {
            router.refresh()
        })

        return () => {
            subscription.unsubscribe()
        }
    }, [router, supabase])

    return (
        <Context.Provider value={{ supabase }}>
            <>{children}</>
        </Context.Provider>
    )
}

export const useSupabase = () => {
    const context = useContext(Context)

    if (context === undefined) {
        throw new Error('useSupabase must be used inside SupabaseProvider')
    }

    return context
}
```

[I read this one before](https://www.reddit.com/r/Supabase/comments/yhf2bt/user_gets_logged_in_automatically_when_resetting/)

**Comment by emrecoban** [comment]
I have fixed this, but if the user refreshes the page after clicking the link, log-in automatically again. Look at [the GIF](https://i.ibb.co/jLLY5Y5/Recording-2023-05-08-at-21-41-14.gif).

User console after clicking the reset link:
![console-output](https://i.ibb.co/qJHG6w4/Screenshot-from-2023-05-08-22-17-55.png)

Updated SupabaseProvider:
```javascript
'use client'

import { createContext, useContext, useEffect, useState } from 'react'
import { createBrowserSupabaseClient } from '@supabase/auth-helpers-nextjs'
import { useRouter } from 'next/navigation'

const Context = createContext(undefined)

export default function SupabaseProvider({ children }) {
    const [supabase] = useState(() => createBrowserSupabaseClient())
    const router = useRouter()

    useEffect(() => {
        const {
            data: { subscription },
        } = supabase.auth.onAuthStateChange((event, session) => {
            if (event === 'PASSWORD_RECOVERY') {
                router.push('/admin/auth/reset');
            }
            if (session && (event === 'SIGNED_IN' || event === 'TOKEN_REFRESHED')) {
                router.refresh()
            }
            if (event === 'SIGNED_OUT') {
                router.refresh()
            }

        })

        return () => {
            subscription.unsubscribe()
        }
    }, [router, supabase])

    return (
        <Context.Provider value={{ supabase }}>
            <>{children}</>
        </Context.Provider>
    )
}

export const useSupabase = () => {
    const context = useContext(Context)

    if (context === undefined) {
        throw new Error('useSupabase must be used inside SupabaseProvider')
    }

    return context
}
```

**Comment by silentworks** [comment]
Hey @emrecoban sorry to hear you are having this issue. I generally advice users away from using the events for password reset stuff, since the `.resetPasswordForEmail` function takes a `redirectTo` URL, you can use that as the direct URL the user should be returned to instead of checking for an event.

```ts
await supabase.auth.resetPasswordForEmail(email, { redirectTo: 'http://example.com/admin/auth/reset' })
```

I've just noticed in your code you are always redirecting the user to `/admin` if they have a session, so no matter which page you go to inside of your logged in area, if you refresh you will be redirected back to `/admin`. This is the cause of your current refreshing issue.

Line 10 - 12 in the first snippet you shared:
```ts
if (data.session) {
    redirect('/admin')
}
```

Note we are releasing a guide for Password reset later on today.

**Reply by emrecoban** [reply]
I'm glad you noticed that. Indeed, the problem was in the 'data.session' if-statement in layout.js. However, I added this statement to prevent users from seeing the login, register, and reset forms if they are already logged in. So if I remove it, it will cause other problems.

```javascript
if (data.session) {
    redirect('/admin')
}
```
I am eagerly waiting for the guide you will release. Where can I follow it? I really need a solution.

**Reply by silentworks** [reply, marked as answer]
You can take a look at this example app using NextJS and also doesn't allow the user to see the login, register and reset form page after they have logged in. Your update password form should be in the logged in area and not in the auth area as the user has to be logged in  to change their password.

https://github.com/supabase-community/supabase-by-example/tree/next/reset-flow/nextjs

**Reply by emrecoban** [reply]
> Your update password form should be in the logged in area and not in the auth area as the user has to be logged in to change their password.

Yes, for now this seems to be the exact solution to the problem. Thank you for your attention.
